stellargraph/random.py

```diff
@@ -75,3 +76,31 @@ def set_seed(seed):
         _rs = _global_state()
     else:
         _rs = _seeded_state(seed)
+
+
+class SeededPerBatch:
+    """
+    Internal utility class for managing a random state per batch number in a multi-threaded
+    environment.
+
+    """
+
+    def __init__(self, create_with_seed, seed):
+        self._create_with_seed = create_with_seed
+        self._walkers = []
+        self._lock = threading.Lock()
+        self._rs, _ = random_state(seed)
+
+    def __getitem__(self, batch_num):
+        self._lock.acquire()
+        try:
+            return self._walkers[batch_num]
+        except IndexError:
+            # always create a new seeded sampler in ascending order of batch number
+            # this ensures seeds are deterministic even when batches are run in parallel
+            for n in range(len(self._walkers), batch_num + 1):
```

Comment:
This could be

```python
self._walks.extend(
    self.create_with_seed(self._rs.randrange(2 ** 32))
    for _ in range(len(self._walkers), batch_num + 1)
)
```
if you thought that was clearer